### PR TITLE
Refactor MSI code to make it easier to add a WiX EXE installer

### PR DESCRIPTION
**Pull request by TravisEz13**
# PR Summary

Refactor MSI generation code to move candle and light to their own function

## PR Context

Refactor msi code to make it easier to add a WiX EXE installer

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by TravisEz13** [comment]
cc @heaths

**Review by adityapatwardhan — APPROVED** [review]

File: tools/packaging/packaging.psm1 [hunk 1]
```diff
@@ -3090,6 +3075,77 @@ function New-MSIPackage
...
+    $extensionArgs = @()
+    foreach ($extensionName in $Extension) {
+        $extensionArgs += '-ext'
+        $extensionArgs += $extensionName
+    }
+
+    $buildArguments = @()
+    foreach ($key in $Argument.Keys) {
+        $buildArguments += "-d$key=`"$($Argument.$key)`""
+    }
```
Comment (iSazonov) [inline tools/packaging/packaging.psm1:3114]: Is it the same as New-MsiArgsArray?

File: tools/packaging/packaging.psm1 [hunk 2]
```diff
@@ -3090,6 +3075,77 @@ function New-MSIPackage
...
+
+    return $buildArguments
+}
+
+function Start-MsiBuild {
+    param(
+        [string[]] $WxsFile,
+        [string[]] $Extension = @('WixUIExtension', 'WixUtilExtension'),
```
Comment (heaths) [inline tools/packaging/packaging.psm1:3094]: If you're gearing up to create a bundle, you may as well as the WixBalExtension here as well. Even if you don't use it now, it won't cause any harm (like linking a .obj file in VC++ you don't use).

File: tools/packaging/packaging.psm1 [hunk 3]
```diff
@@ -3090,6 +3075,77 @@ function New-MSIPackage
...
+
+    $resolvedWxsFiles = @()
+    foreach ($file in $WxsFile) {
+        $resolvedWxsFiles += (Resolve-Path -Path $file).ProviderPath
+    }
+
+    Write-Verbose "$resolvedWxsFiles" -Verbose
+
+    Write-Log "running candle..."
+    Start-NativeExecution -VerboseOutputOnError { & $wixPaths.wixCandleExePath $resolvedWxsFiles -out "$outDir\\" $extensionArgs -arch $ProductTargetArchitecture $buildArguments -v}
```
Comment (heaths) [inline tools/packaging/packaging.psm1:3135]: Is the intent here to use msbuild eventually, or just to build MSIs and/or bundles separately?

**Review by heaths — APPROVED** [review]
Nice clean-up!

File: tools/packaging/packaging.psm1 (hunk 2 above)
Comment (TravisEz13) [reply tools/packaging/packaging.psm1:3094]:
```suggestion
        [string[]] $Extension = @('WixUIExtension', 'WixUtilExtension', 'WixBalExtension'),
```

File: tools/packaging/packaging.psm1 (hunk 3 above)
Comment (TravisEz13) [reply tools/packaging/packaging.psm1:3135]: separately.

**Pull request merged**